quickwit/quickwit-doc-mapper/src/default_doc_mapper/date_time_format.rs

```diff
@@ -0,0 +1,159 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::convert::Infallible;
+use std::fmt::Display;
+use std::str::FromStr;
+
+use serde::{Deserialize, Deserializer, Serialize};
+use unwrap_infallible::UnwrapInfallible;
+
+/// Specifies the datetime and unix timestamp formats to use when parsing date strings.
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub enum DateTimeFormat {
+    ISO8601,
+    RFC2822,
+    RCF3339,
+    Strftime {
+        strftime_format: String,
+        with_timezone: bool,
+    },
+    TimestampSecs,
+    TimestampMillis,
+    TimestampMicros,
+}
+
+impl DateTimeFormat {
+    pub fn as_str(&self) -> &str {
+        match self {
+            DateTimeFormat::ISO8601 => "iso8601",
+            DateTimeFormat::RFC2822 => "rfc2822",
+            DateTimeFormat::RCF3339 => "rfc3339",
+            DateTimeFormat::Strftime {
+                strftime_format, ..
+            } => strftime_format,
+            DateTimeFormat::TimestampSecs => "unix_ts_secs",
+            DateTimeFormat::TimestampMillis => "unix_ts_millis",
+            DateTimeFormat::TimestampMicros => "unix_ts_micros",
+        }
+    }
+
+    pub fn is_timestamp(&self) -> bool {
+        matches!(
+            self,
+            DateTimeFormat::TimestampSecs
+                | DateTimeFormat::TimestampMillis
+                | DateTimeFormat::TimestampMicros
+        )
+    }
+}
+
+impl Display for DateTimeFormat {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str(self.as_str())
+    }
+}
+
+impl FromStr for DateTimeFormat {
+    type Err = Infallible;
+
+    fn from_str(date_time_format_str: &str) -> Result<Self, Self::Err> {
+        let date_time_format = match date_time_format_str.to_lowercase().as_str() {
+            "iso8601" => DateTimeFormat::ISO8601,
+            "rfc2822" => DateTimeFormat::RFC2822,
+            "rfc3339" => DateTimeFormat::RCF3339,
+            "unix_ts_secs" => DateTimeFormat::TimestampSecs,
+            "unix_ts_millis" => DateTimeFormat::TimestampMillis,
+            "unix_ts_micros" => DateTimeFormat::TimestampMicros,
+            _ => DateTimeFormat::Strftime {
+                strftime_format: date_time_format_str.to_string(),
+                with_timezone: date_time_format_str.contains("%z"),
+            },
+        };
+        Ok(date_time_format)
+    }
+}
+
+impl Serialize for DateTimeFormat {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where S: serde::Serializer {
+        serializer.serialize_str(self.as_str())
+    }
+}
+
+impl<'de> Deserialize<'de> for DateTimeFormat {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where D: Deserializer<'de> {
+        let date_time_format_str = String::deserialize(deserializer)?;
+        Ok(DateTimeFormat::from_str(&date_time_format_str).unwrap_infallible())
```

Comment:
TIL I learnt about `unwrap_infallible`

Reply:
Courtesy of @evanxg852000

Reply:
Yes, I saw it yesterday, indeed very nice to learn this.
Aside I just could not figure out how @fulmicoton came up with these numbers, I tried understanding the bit manipulation required Thanks we have a issue coming up on that.
```
28-33 ---> that's seconds
38-43 ---> that's millis
48-53 ---> that's micros
58-63 ---> that's nanos
```